src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -0,0 +1,475 @@
+
+        Returns:
+            High level metric calculated from the given metrics.
+            Return None if there are no component metrics.
+        """
+        # In the future we will have lower bound, too.
+        consume_bounds = Bound(0, 0)
+        supply_bounds = Bound(0, 0)
+        latest_timestamp: datetime = _MIN_TIMESTAMP
+        any_metrics = False
+
+        for bid in working_batteries:
+            consume_upper_bounds: List[float] = []
+            supply_upper_bounds: List[float] = []
+
+            if not self._is_missing(bid, metrics_data):
+                data = metrics_data[bid]
+                latest_timestamp = data.timestamp
+                if data.has_metrics():
+                    consume_upper_bounds.append(
+                        data.get(ComponentMetricId.POWER_UPPER_BOUND)
+                    )
+                    supply_upper_bounds.append(
+                        data.get(ComponentMetricId.POWER_LOWER_BOUND)
+                    )
+
+            inverter_id = self._bat_inv_map[bid]
+            if not self._is_missing(inverter_id, metrics_data):
+                data = metrics_data[inverter_id]
+                latest_timestamp = data.timestamp
+                if data.has_metrics():
+                    consume_upper_bounds.append(
+                        data.get(ComponentMetricId.ACTIVE_POWER_UPPER_BOUND)
+                    )
+                    supply_upper_bounds.append(
+                        data.get(ComponentMetricId.ACTIVE_POWER_LOWER_BOUND)
+                    )
+
+            if len(consume_upper_bounds) > 0:
+                any_metrics = True
+                consume_bounds.upper += min(consume_upper_bounds)
+                supply_bounds.upper += min(supply_upper_bounds)
+
+        if latest_timestamp == _MIN_TIMESTAMP:
+            _logger.warning("Can't calculate formula, no data available.")
+            return None
+
+        power_metrics = (
+            PowerBoundsMetrics(consume_bounds, supply_bounds) if any_metrics else None
+        )
+
+        return PowerBounds(
+            timestamp=latest_timestamp,
+            metrics=power_metrics,
+        )
+
+
+class ActivePowerFormula(MetricCalculator[Sample]):
+    """Define how to calculate active power metrics."""
+
+    def __init__(
+        self,
+        batteries: Set[int],
+    ) -> None:
+        """Create class instance.
+
+        Args:
+            batteries: What batteries should be used for calculation.
+        """
+        super().__init__(batteries)
+
+        self._bat_inv_map = battery_utils.battery_inverter_mapping(batteries)
+
+        if len(self._bat_inv_map) == 0:
+            raise ValueError(
+                f"Formula {PowerBoundsCalculator.name} requires batteries "
+                "with adjacent inverter",
+            )
+
+        if len(batteries) != len(self._bat_inv_map):
+            _logger.warning(
+                "Some batteries hasn't adjacent inverter. Use %s out of %s",
```

Comment:
This is a 3rd warning for the missing inverter, you **really hate** missing inverters :laughing:

Reply:
Hmm I don't hate them :D  

It is just still unknown how much inverters battery can have.
First we could have ONLY one inverter per one battery.
Now it can have ZERO or ONE inverter per one battery
In the future we plan to have [0-inf] inverters per one battery. 

But you are right . There are to much logs.  I simplified them the logs a lot.